src/lib.rs

```diff
@@ -1,47 +1,78 @@
 #![no_std]
+use bit_field::BitField;
+use embedded_hal::blocking::i2c::{Write, WriteRead};
+use num_enum::IntoPrimitive;
+use paste::paste;
 
-//! Manages an MCP23017, a 16-Bit I2C I/O Expander with Serial Interface module.
-//!
-//! This operates the chip in `IOCON.BANK=0` mode, i.e. the registers are mapped sequentially.
-//! This driver does not set `IOCON.BANK`, but the factory default is `0` and this driver does
-//! not change that value.
-//!
-//! See [the datasheet](http://ww1.microchip.com/downloads/en/DeviceDoc/20001952C.pdf) for more
-//! information on the device.
+/// Pin modes.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Direction {
+    /// Represents input mode.
+    Input,
+    /// Represents output mode.
+    Output,
+}
 
-#![deny(
-    missing_docs,
-    missing_debug_implementations,
-    missing_copy_implementations,
-    trivial_casts,
-    trivial_numeric_casts,
-    unstable_features,
-    unused_import_braces,
-    unused_qualifications,
-    warnings
-)]
-#![allow(dead_code, non_camel_case_types)]
-#![allow(clippy::uninit_assumed_init, clippy::upper_case_acronyms)]
+/// Pin levels.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Level {
+    /// High level
+    High,
+    /// Low level
+    Low,
+}
 
-extern crate embedded_hal as ehal;
+/// Pin Pull Up state.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum PullUp {
+    /// Weak pull up enabled
+    Enabled,
+    /// Weak pull up disabled, pin floating
+    Disabled,
+}
 
-use ehal::blocking::i2c::{Write, WriteRead};
+/// Interrupt on change state.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum IntOnChange {
+    /// Enabled
+    Enabled,
+    /// Disables
+    Disabled,
+}
 
-/// The default I2C address of the MCP23017.
-const DEFAULT_ADDRESS: u8 = 0x20;
+/// Interrupt control.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum IntMode {
+    /// Interrupt on level (seel DEFVAL, )
+    OnLevel,
+    /// Interrupt on change (see GPINTEN, )
+    OnChange,
+}
 
-/// Binary constants.
-const HIGH: bool = true;
-const LOW: bool = false;
+/// Interrupt flag.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum IntFlag {
```

Comment:
Why not just use `bool` to indicate asserted?

Reply:
Here it seems a bit clearer that a `true` is an asserted interrupt.

Reply:
But it doesn't work well with the macro. I'll keep it and since this is an RO register (writes ignored) will just `repr(bool)` that allows convenient conversion.

Reply:
Ah. `repr(bool)` doesn't exist.